crates/circuit/src/vf2.rs

```diff
@@ -695,494 +839,544 @@ where
+                || (self.directed() && !self.mapped_edges_counts_match(nodes, Incoming))
+            {
+                return Ok(None);
             }
-            let val = edge_multiplicity(&st.1.graph, &st.1.adjacency_matrix, nodes[1], n_neigh);
+            S::Score::id()
+        };
 
-            let has_edge = is_adjacent(&st.0.graph, &st.0.adjacency_matrix, nodes[0], m_neigh, val);
-            if !has_edge {
-                return Ok(false);
+        let new_score = S::Score::combine(
+            &S::Score::combine(&self.semantic.cur, &node_score),
+            &edge_score,
+        );
+        match self.semantic.limit.as_ref() {
+            Some(limit) => {
+                // We're not consistent if the score  breaks the limit.
+                Ok((S::Score::cmp(&new_score, limit) != Ordering::Greater).then_some(new_score))
             }
+            None => Ok(Some(new_score)),
         }
-        if succ_count[0].cmp(&succ_count[1]).then(ordering) != ordering {
-            return Ok(false);
-        }
-        // R_pred
-        if st.0.graph.is_directed() {
-            let mut pred_count = [0, 0];
-            for n_neigh in st.0.graph.neighbors_directed(nodes[0], Incoming) {
-                pred_count[0] += 1;
-                if !induced {
-                    continue;
-                }
-                // the self loop case is handled in outgoing
-                let m_neigh = st.0.mapping[n_neigh.index()];
-                if m_neigh == end {
-                    continue;
-                }
-                let val = edge_multiplicity(&st.0.graph, &st.0.adjacency_matrix, n_neigh, nodes[0]);
+    }
 
-                let has_edge =
-                    is_adjacent(&st.1.graph, &st.1.adjacency_matrix, m_neigh, nodes[1], val);
-                if !has_edge {
-                    return Ok(false);
+    /// Are the directed edges of these two paired nodes consistent with all nodes already in the
+    /// mapping?
+    ///
+    /// This is the more expensive semantic check, where we have to run pairwise through the edge
+    /// lists to match them up against each other.  If there's no semantic checking enabled, use
+    /// `mapped_edges_counts_match`, which short-circuits based on the pre-calculated adjacency
+    /// matrices.
+    fn mapped_edges_semantic_match(
+        &self,
+        nodes: [NodeIndex; 2],
+        direction: Direction,
+    ) -> Result<Option<S::Score>, S::EdgeError> {
+        let needle = &self.st.0;
+        let haystack = &self.st.1;
+        // We only handle self loops in the `Outgoing` direction to avoid double-counting the score.
+        let handle_self = direction == Direction::Outgoing;
+
+        macro_rules! neighbor {
+            ($edge:expr) => {
+                match direction {
+                    Direction::Outgoing => $edge.target(),
+                    Direction::Incoming => $edge.source(),
                 }
-            }
+            };
+        }
```

Comment:
Why is this implemented as a macro?

Reply:
The type of an "edge" between the needle graph and the haystack graph can be different, and the purpose of this macro is to provide a closure over the `direction` argument.  Rust closures can't take generic arguments, and inline `fn` can't close over values, so for this very simple closure, I'd thought it was a little easier to have it as a macro.